test/system/src/test/java/io/pravega/test/system/WatermarkingTest.java

```diff
@@ -0,0 +1,287 @@
+        EventStreamClientFactory clientFactory = EventStreamClientFactory.withScope(SCOPE, clientConfig);
+        JavaSerializer<Long> javaSerializer = new JavaSerializer<>();
+        @Cleanup
+        EventStreamWriter<Long> writer1 = clientFactory.createEventWriter(STREAM, javaSerializer,
+                EventWriterConfig.builder().build());
+        @Cleanup
+        EventStreamWriter<Long> writer2 = clientFactory.createEventWriter(STREAM, javaSerializer,
+                EventWriterConfig.builder().build());
+
+        AtomicBoolean stopFlag = new AtomicBoolean(false);
+        // write events
+        writeEvents(writer1, stopFlag);
+        writeEvents(writer2, stopFlag);
+
+        // scale the stream several times so that we get complex positions
+        Stream streamObj = Stream.of(SCOPE, STREAM);
+        scale(controller, streamObj);
+
+        // wait until mark stream is created
+        AtomicBoolean markStreamCreated = new AtomicBoolean(false);
+        Futures.loop(() -> !markStreamCreated.get(), 
+                () -> Futures.exceptionallyExpecting(controller.getCurrentSegments(SCOPE, NameUtils.getMarkStreamForStream(STREAM)), 
+                    e -> Exceptions.unwrap(e) instanceof StoreException.DataNotFoundException, null)
+                .thenAccept(v -> markStreamCreated.set(v != null)), executorService);
+
+        @Cleanup
+        SynchronizerClientFactory syncClientFactory = SynchronizerClientFactory.withScope(SCOPE, clientConfig);
+        String markStream = NameUtils.getMarkStreamForStream(STREAM);
+
+        RevisionedStreamClient<Watermark> watermarkReader = syncClientFactory.createRevisionedStreamClient(markStream,
+                new WatermarkSerializer(),
+                SynchronizerConfig.builder().build());
+        
+        LinkedBlockingQueue<Watermark> watermarks = new LinkedBlockingQueue<>();
+        fetchWatermarks(watermarkReader, watermarks, stopFlag);
+
+        AssertExtensions.assertEventuallyEquals(true, () -> watermarks.size() >= 2, 100000);
+
+        // scale down one controller instance. 
+        Futures.getAndHandleExceptions(controllerInstance.scaleService(1), ExecutionException::new);
+
+        // wait until at least 2 more watermarks are emitted
+        AssertExtensions.assertEventuallyEquals(true, () -> watermarks.size() >= 4, 100000);
+
+        stopFlag.set(true);
+
+        Watermark watermark0 = watermarks.take();
+        Watermark watermark1 = watermarks.take();
+        Watermark watermark2 = watermarks.take();
+        Watermark watermark3 = watermarks.take();
+        
+        assertTrue(watermark0.getLowerTimeBound() < watermark0.getUpperTimeBound());
+        assertTrue(watermark1.getLowerTimeBound() < watermark1.getUpperTimeBound());
+        assertTrue(watermark2.getLowerTimeBound() < watermark2.getUpperTimeBound());
+        assertTrue(watermark3.getLowerTimeBound() < watermark3.getUpperTimeBound());
+
+        // verify that watermarks are increasing in time.
```

Comment:
The upper time bounds should also go up.

Reply:
that may not be true necessarily, at least in the current implementation. 
if a writer had reported a higher time in previous watermark computation and is no longer participating in the subsequent watermarking computation then the upper bound (which is upper bound on all reported times) could be lower than a previously reported high time. 

We could set it to max(previous.upperbound, currentUpperBound) which will guarantee that its always increasing.

Reply:
ok